scapy/layers/inet.py

```diff
@@ -1181,86 +1152,112 @@ def overlap_frag(p, overlap, fragsize=8, overlap_fragsize=None):
+class BadFragments(ValueError):
+    def __init__(self, *args, **kwargs):
+        self.frags = kwargs.pop("frags", None)
+        super(BadFragments, self).__init__(*args, **kwargs)
+
+
+def _defrag_iter_and_check_offsets(frags):
+    """
+    Internal generator used in _defrag_ip_pkt
+    """
+    offset = 0
+    for pkt, o, length in frags:
+        if offset != o:
+            if offset > o:
+                warning("Fragment overlap (%i > %i) on %r" % (offset, o, pkt))
+            raise BadFragments
+        offset += length
+        yield bytes(pkt[IP].payload)
+
+
+def _defrag_ip_pkt(pkt, frags):
+    """
+    Defragment a single IP packet.
+
+    :param pkt: the new pkt
+    :param frags: a defaultdict(list) used for storage
+    :return: a tuple (fragmented, defragmented_value)
+    """
+    ip = pkt[IP]
+    if pkt.frag != 0 or ip.flags.MF:
+        # fragmented !
+        uid = (ip.id, ip.src, ip.dst, ip.proto)
+        if ip.len is None or ip.ihl is None:
+            fraglen = len(ip.payload)
         else:
-            c_len += q[IP].len - (q[IP].ihl << 2)
-        if conf.padding_layer in q:
-            del q[conf.padding_layer].underlayer.payload
-        txt.add_payload(q[IP].payload.copy())
-        if q.time > p.time:
-            p.time = q.time
-    else:
-        ip.flags.MF = False
-        del ip.chksum
-        del ip.len
-        p = p / txt
-        p._defrag_pos = max(x._defrag_pos for x in lst)
-        defrag.append(p)
+            fraglen = ip.len - (ip.ihl << 2)
+        # (pkt, frag offset, frag len)
+        frags[uid].append((pkt, ip.frag << 3, fraglen))
+        if not ip.flags.MF:  # no more fragments = last fragment
+            curfrags = sorted(frags[uid], key=lambda x: x[1])  # sort by offset
+            try:
+                data = b"".join(_defrag_iter_and_check_offsets(curfrags))
+            except ValueError:
+                # bad fragment
+                badfrags = frags[uid]
+                del frags[uid]
+                raise BadFragments(frags=badfrags)
+            # re-build initial packet without fragmentation
+            p = curfrags[0][0].copy()
+            pay_class = p[IP].payload.__class__
+            p[IP].flags.MF = False
+            p[IP].remove_payload()
+            p[IP].len = None
+            p[IP].chksum = None
+            # append defragmented payload
+            p /= pay_class(data)
+            # cleanup
+            del frags[uid]
+            return True, p
+        return True, None
+    return False, pkt
 
 
 def _defrag_logic(plist, complete=False):
-    """Internal function used to defragment a list of packets.
+    """
+    Internal function used to defragment a list of packets.
     It contains the logic behind the defrag() and defragment() functions
     """
-    frags = defaultdict(lambda: [])
+    frags = defaultdict(list)
     final = []
-    pos = 0
-    for p in plist:
-        p._defrag_pos = pos
-        pos += 1
-        if IP in p:
```

Comment:
Looks like without this check `sniff` fails with
```python
Traceback (most recent call last):
    sniff(offline='pcap-with-arp-stuff.pcap', session=IPSession)
  File "scapy/sendrecv.py", line 1311, in sniff
    sniffer._run(*args, **kwargs)
  File "scapy/scapy/sendrecv.py", line 1254, in _run
    session.on_packet_received(p)
  File "scapy/scapy/sessions.py", line 144, in on_packet_received
    super(IPSession, self).on_packet_received(self._ip_process_packet(pkt))
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "scapy/scapy/sessions.py", line 138, in _ip_process_packet
    return _defrag_ip_pkt(packet, self.fragments)[1]  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "scapy/scapy/layers/inet.py", line 1183, in _defrag_ip_pkt
    ip = pkt[IP]
         ~~~^^^^
  File "scapy/scapy/packet.py", line 1370, in __getitem__
    raise IndexError("Layer [%s] not found" % name)
IndexError: Layer [IP] not found
```
when pcaps contain at least one ARP packet (or any other packet without the IP layer).

Reply:
Good catch, thanks !